### DISCUSSION TITLE
Constrained shortest path algorithm

### DISCUSSION THREAD

**Discussion by lhoupert** [networkx/networkx#4832] [Algorithms]
Hi, 

I am using networkx to work on shortest path problems. I am mostly using [shortest_path](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.shortest_paths.generic.shortest_path.html#networkx.algorithms.shortest_paths.generic.shortest_path). I was wondering if, with the current version of networkx, it is possible to constrain the shortest path calculation?

For example, I would like to compute the shortest path between two nodes based on a _'length'_ attribute but I also want to impose that the sum of the _'height'_ attribute along the path is inferior to a specific threshold?

Is there a way to do that with networkx? 

I opened a stackoverflow post about this here: https://stackoverflow.com/questions/67674893/constrained-short-path-algorithm-in-networkx

**Comment by MridulS** [comment]
A not-so-efficient way of this could be to compute all the paths and then go through them to find the shortest path which satisfies the condition, something like [NOT TESTED!]
``` python
 current_min = float('inf')
 current_path = None
 for path in nx.all_simple_paths(G, source, target):
     if nx.path_weight(G, path, 'height') < threshold:
         if nx.path_weight(G, path, 'weight') < current_min:
             current_min = nx.path_weight(G, path, 'weight')
             current_path = path
 print(current_path, current_min)
```
This will be painfully slow for a large graph.

**Reply by lhoupert** [reply]
Thanks @MridulS , I would like to find a solution where I am not computing all the paths :-) I will have a lot of data

**Comment by boothby** [comment]
Reading through `_dijkstra_multisource`, we can hack this in.  The trick is to implement a special class that acts numerical...

```python
import networkx as nx
class length_and_height:
    def __init__(self, l, h):
        self.l = l
        self.h = h

    def __add__(self, other):
        if isinstance(other, length_and_height):
            return length_and_height(self.l + other.l, self.h + other.h)
        else:
            return length_and_height(self.l + other, self.h + other)

    def __radd__(self, other):
        if isinstance(other, length_and_height):
            return length_and_height(self.l + other.l, self.h + other.h)
        else:
            return length_and_height(self.l + other, self.h + other)


    def __gt__(self, other):
        if isinstance(other, length_and_height):    
            if self.l is None or other.l is None:
                return self.h > other.h
            else:
                return self.l > other.l
        else:
            return self.l > other

    def __lt__(self, other):
        if isinstance(other, length_and_height):    
            if self.l is None or other.l is None:
                return self.h < other.h
            else:
                return self.l < other.l
        else:
            return self.l < other
            
    def __repr__(self):
        return f'length_and_height({self.l!r}, {self.h!r})'

def weight_function(u, v, d):
    return length_and_height(d['length'], d['height'])

G = nx.Graph()

G.add_edge(0, 1, height=101, length=1)
G.add_edge(0, 2, height=50, length=1)
G.add_edge(2, 1, height=50, length=1)

height_bound = length_and_height(None, 100)
path = nx.single_source_dijkstra(G, 0, 1, weight=weight_function, cutoff=height_bound)
```

However, this is a hack, and won't necessarily get the right answer.  Specifically, if there are two admissably short (in height) paths, the algorithm will find the shorter (in length) one regardless of its height.  This can make it look impossible to find a node, when an admissable path exists.  This is visible by extending the example by one node...

```python
G.add_edge(0, 1, height=100, length=1)
G.add_edge(0, 2, height=40, length=1)
G.add_edge(2, 1, height=40, length=1)
G.add_edge(1, 3, height=20, length=1)

height_bound = length_and_height(None, 100)
path = nx.single_source_dijkstra(G, 0, 3, weight=weight_function, cutoff=height_bound) #raises NetworkXNoPath: No path to 3.

```

**Reply by lhoupert** [reply]
Thanks @boothby, it is a good start. I will have to think how I could use this.... 

In your example I can find a path if I swap the target node by the source node:
``` python
path = nx.single_source_dijkstra(G, 3, 0, weight=weight_function, cutoff=height_bound) #raises NetworkXNoPath: No path to 3.
print(path)
``` 
```
(length_and_height(3, 100), [3, 1, 2, 0])
```
It is a shame that there is no equivalent function `single_target_dijkstra` (similar to [[this](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.shortest_paths.unweighted.single_target_shortest_path.html#networkx.algorithms.shortest_paths.unweighted.single_target_shortest_path)] but for weighted graph). 

Do you think it worth developing such a function? Or maybe just using the [DiGraph.reverse function](https://networkx.org/documentation/stable/reference/classes/generated/networkx.DiGraph.reverse.html) would do the job?

If I defined a directed graph to start with, I can run this bit of code to find the correct reverse path:
``` python
G = nx.DiGraph()
G.add_edge(0, 1, height=100, length=1)
G.add_edge(0, 2, height=40, length=1)
G.add_edge(2, 1, height=40, length=1)
G.add_edge(1, 3, height=20, length=1)

height_bound = length_and_height(None, 100)
G_reverse = nx.DiGraph.reverse(G)
path_reverse = nx.single_source_dijkstra(G_reverse, 3, 0, weight=weight_function, cutoff=height_bound) 
print(f"reverse path: {path_reverse}")
```
```
reverse path: (length_and_height(3, 100), [3, 1, 2, 0])
```

**Reply by dschult** [reply]
Yes, to get single target, reverse the graph and use single source. :}

**Reply by boothby** [reply]
> In your example I can find a path if I swap the target node by the source node:

Please note that this doesn't get you around the problem with the algorithm, it only works on that one example.  After tinkering with it for a while, I've convinced myself that this is probably a hard problem in general (if it's in NP, I don't know what a polytime-verifiable certificate would look like), and a modified version of Lawler-Yen might be the best approach to take.

https://en.wikipedia.org/wiki/Yen%27s_algorithm

**Reply by boothby** [reply]
Here's a symmetric example that can't be fixed by reversing:
```python
G = nx.Graph()
G.add_edge(0, 1, height=100, length=1)
G.add_edge(0, 2, height=40, length=1)
G.add_edge(2, 1, height=40, length=1)
G.add_edge(1, 3, height=20, length=1)
G.add_edge(3, 4, height=20, length=1)
G.add_edge(4, 5, height=40, length=1)
G.add_edge(5, 6, height=40, length=1)
G.add_edge(4, 6, height=100, length=1)

height_bound = length_and_height(None, 200)
path = nx.single_source_dijkstra(G, 0, 6, weight=weight_function, cutoff=height_bound)
```

**Reply by lhoupert** [reply]
Thank you for tinkering with it! I thought also about using the Yen approach, particularly because I would like to use this approach for negative weights :-) (see other discussion I opened [here](https://github.com/networkx/networkx/discussions/4835)) 

It seems that the Yen's algorithm is already implemented in the function [shortest_simple_paths](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.simple_paths.shortest_simple_paths.html#networkx.algorithms.simple_paths.shortest_simple_paths). However looking at the [source code](https://networkx.org/documentation/stable/_modules/networkx/algorithms/simple_paths.html#shortest_simple_paths) is `weight is not None` it's `_bidirectional_dijkstra` which is used...

Do you think it would be worth trying to implement a cutoff parameter in `shortest_simple_paths` working in a similar way that in `_dijkstra_multisource`?

**Comment by SPA20** [comment]
Hi,

Thanks for the discussion. Just wondering whether there is any progress on ECMP(Equal Cost Multi-Path) feature in SPF algorithms. I am hitting the same issue with the method selected as'dijkstra' . My graph has parallel paths with same cost towards destination. However, shortest picks only one path out of multiple available options. Any other alternative python modules which can do this feature?

Thanks.

**Comment by amcandio** [comment]
@lhoupert a bit late here but I recently wrote a blog post about this. See trick number 2 in https://blog.scientific-python.org/networkx/hacking-shortest-paths/.

To solve this problem you need to model the graph so each node becomes a state rather than just a location. So each node becomes a tuple (location, current height). Your edges now become transitions between states, so you need to make sure you transition to the right height when you take an edge.

You can skip the nodes where the height is higher than > 300. This way you encode your constraint in the graph itself.

Last, you can use Dijkstra to compute the shortest path from your source (actually, from (source, initial height)) to all nodes of the form (destination, some height). Since you have multiple targets, you can use trick number 1 in the blogpost.

For that trick you can add a sentinel/sync node that has connections from all targets. Then you can use regular Dijkstra's algorithm to compute the shortest path from your source to the dummy target.

Note that, as mentioned in the post, this approach can lead to a graph explosion if you have many possible states. This is expected because the problem does not have a polynomial solution. To mitigate that you can use some pruning techniques to reduce the amount of different states. But then solution becomes an approximation.
